File: common/src/main/java/dev/number6/comprehend/dagger/ComprehensionServiceModule.java
package dev.number6.comprehend.dagger;

import dev.number6.comprehend.AwsComprehendClient;
import dev.number6.comprehend.port.ComprehensionPort;
import dev.number6.comprehend.adaptor.AwsComprehensionAdaptor;
import com.amazonaws.services.comprehend.AmazonComprehend;
import com.amazonaws.services.comprehend.AmazonComprehendClientBuilder;
import dagger.Module;
import dagger.Provides;

@Module
public class ComprehensionServiceModule {

    @Provides
    public ComprehensionPort providesComprehensionPort(AwsComprehendClient client) {
        return new AwsComprehensionAdaptor(client);
    }

    @Provides
    public AwsComprehendClient providesComprehensionClient(AmazonComprehend amazonComprehend) {
        return new AwsComprehendClient(amazonComprehend);
    }

    @Provides
    public AmazonComprehend providesAmazonComprehend() {
        return AmazonComprehendClientBuilder.defaultClient();
    }
}

File: common/src/main/kotlin/dev/number6/comprehend/dagger/ComprehensionServiceModule.kt
package dev.number6.comprehend.dagger

import com.amazonaws.services.comprehend.AmazonComprehend
import com.amazonaws.services.comprehend.AmazonComprehendClientBuilder
import dagger.Module
import dagger.Provides
import dev.number6.comprehend.AwsComprehendClient
import dev.number6.comprehend.adaptor.AwsComprehensionAdaptor
import dev.number6.comprehend.port.ComprehensionPort
import javax.inject.Singleton

@Module
class ComprehensionServiceModule {
    @Provides
    @Singleton
    fun providesComprehensionPort(client: AwsComprehendClient): ComprehensionPort {
        return AwsComprehensionAdaptor(client)
    }

    @Provides
    @Singleton
    fun providesComprehensionClient(amazonComprehend: AmazonComprehend): AwsComprehendClient {
        return AwsComprehendClient(amazonComprehend)
    }

    @Provides
    @Singleton
    fun providesAmazonComprehend(): AmazonComprehend {
        return AmazonComprehendClientBuilder.defaultClient()
    }
}